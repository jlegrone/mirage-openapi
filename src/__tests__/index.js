import Mirage from 'mirage-config' // eslint-disable-line
import mirageOpenAPI from '../index'
import definition from './data/petstore.swagger-v2.json'

let server
let instance
let routers

const fakeRequest = ({
  method = 'GET',
  requestHeaders = { 'Content-Type': 'application/json' },
  url = 'http://petstore.swagger.io/v2/',
}) => ({
  method,
  requestHeaders,
  url,
})

beforeAll(async () => {
  const configs = [
    { definition, namespace: 'http://petstore.swagger.io/v2' },
    { definition, namespace: '/' },
  ]
  server = new Mirage()
  instance = mirageOpenAPI({ configs, server })
  routers = await instance.resolveRouters()
})

describe('mirage openAPI instance', () => {
  test('generates router', () => {
    expect(routers).toHaveLength(2)
    expect(routers[0].namespace).toEqual('http://petstore.swagger.io/v2')
  })

  test('matches existing route', () => {
    const [{ router }] = routers
    const matches = router.recognize('/pet')
    expect(matches).toMatchSnapshot()
  })

  test('matches existing dynamic route', () => {
    const [{ router }] = routers
    const matches = router.recognize('/pet/0')
    expect(matches).toMatchSnapshot()
  })

  test('does not match missing route', () => {
    const [{ router }] = routers
    const matches = router.recognize('/foo')
    expect(matches).toEqual(undefined)
  })

  test('throws error for missing namespace', () => {
    const response = instance.handler({}, fakeRequest({ url: 'http://example.com' }))
    return expect(response).rejects.toMatchSnapshot()
  })

  test('throws error for missing route', () => {
    const response = instance.handler({}, fakeRequest({ url: 'http://petstore.swagger.io/v2/foo' }))
    return expect(response).rejects.toMatchSnapshot()
  })

  test('throws error for missing route method', () => {
    const response = instance.handler({}, fakeRequest({ url: 'http://petstore.swagger.io/v2/pet' }))
    return expect(response).rejects.toMatchSnapshot()
  })

  test('responds with example data', async () => {
    const response = await instance.handler(
      {},
      fakeRequest({ url: 'http://petstore.swagger.io/v2/store/inventory' }),
    )

    expect(response.headers['Content-Type']).toBe('application/json')
    expect(response.code).toBe(200)
    expect(response.data).toMatchSnapshot()
  })

  test('responds with sample data', async () => {
    const response = await instance.handler(
      {},
      fakeRequest({ url: 'http://petstore.swagger.io/v2/pet/1' }),
    )

    expect(response.headers['Content-Type']).toBe('application/json')
    expect(response.code).toBe(200)
    expect(response).toHaveProperty('data')
    expect(response.data).toHaveProperty('name')
  })

  test('validates bad requests', async () => {
    const response = await instance.handler(
      {},
      fakeRequest({ url: 'http://petstore.swagger.io/v2/pet', method: 'POST' }),
    )

    expect(response.headers['Content-Type']).toBe('application/json')
    expect(response.code).toBe(400)
    expect(response.data).toMatchSnapshot()
  })
})
