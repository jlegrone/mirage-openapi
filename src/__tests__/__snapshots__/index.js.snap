// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mirage openAPI instance matches existing dynamic route 1`] = `
RecognizeResults {
  "0": Object {
    "handler": Object {
      "delete": [Function],
      "get": [Function],
      "post": [Function],
    },
    "isDynamic": true,
    "params": Object {
      "petId": "0",
    },
  },
  "length": 1,
  "queryParams": Object {},
}
`;

exports[`mirage openAPI instance matches existing route 1`] = `
RecognizeResults {
  "0": Object {
    "handler": Object {
      "post": [Function],
      "put": [Function],
    },
    "isDynamic": false,
    "params": Object {},
  },
  "length": 1,
  "queryParams": Object {},
}
`;

exports[`mirage openAPI instance responds with example data 1`] = `
"{
  \\"cats\\": 42,
  \\"dogs\\": 9,
  \\"lizards\\": 1201
}"
`;

exports[`mirage openAPI instance throws error for missing namespace 1`] = `
[Error: MirageOpenAPI: There is no route defined which matches http://example.com.
This url does not match any existing namespaces.]
`;

exports[`mirage openAPI instance throws error for missing route 1`] = `
[Error: MirageOpenAPI: There is no route defined which matches http://petstore.swagger.io/v2/foo.
There is no handler available in your openAPI config for this route and HTTP method.]
`;

exports[`mirage openAPI instance throws error for missing route method 1`] = `[TypeError: handler is not a function]`;

exports[`mirage openAPI instance validates bad requests 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "INVALID_REQUEST_PARAMETER",
      "errors": Array [
        Object {
          "code": "OBJECT_MISSING_REQUIRED_PROPERTY",
          "message": "Missing required property: photoUrls",
          "params": Array [
            "photoUrls",
          ],
          "path": Array [],
        },
        Object {
          "code": "OBJECT_MISSING_REQUIRED_PROPERTY",
          "message": "Missing required property: name",
          "params": Array [
            "name",
          ],
          "path": Array [],
        },
      ],
      "in": "body",
      "message": "Invalid parameter (body): Value failed JSON Schema validation",
      "name": "body",
      "path": Array [
        "paths",
        "/pet",
        "post",
        "parameters",
        "0",
      ],
    },
  ],
  "warnings": Array [],
}
`;
